library/objective-c/EnvoyEngineImpl.m

```diff
@@ -29,13 +115,33 @@ - (void)dealloc {
   [[NSNotificationCenter defaultCenter] removeObserver:self];
 }
 
+- (int)registerFilter:(EnvoyHTTPFilter *)filter {
+  // TODO(goaway): Everything here leaks, but it's all be tied to the life of the engine.
+  // This will need to be updated for https://github.com/lyft/envoy-mobile/issues/332
+  ios_http_filter_context *context = safe_malloc(sizeof(ios_http_filter_context));
```

Comment:
following on @rebello95's comment about lifetime. We definitely want a per stream instance of the full chain, like Envoy. 

I wonder if what we can do is: have the api retain a filter factory in its context, and register that. Then the registry can instantiate a filter chain, that the stream retains, and passes that in the context to the static `ios_http_filter_on_*_*` functions. So we would have two different contexts, and the one here (which retains the factory) leaks, but will eventually be tied/retained to/by the engine instance.

Reply:
That's one of the options that I've been considering, but there's some added complexity with the memory management of the instance. We'll need a dealloc hook as well, and we need to guard against the platform implementation living longer than the underlying core instantiation (and more specifically, trying to interact with it after it's gone).

I've actually been leaning towards a more integrated approach - with the API presenting both an `init()` and a `release()` function, the former of which will populate the rest of the function pointers on the struct, and the latter of which will nullify them (and trigger memory-safe cleanup).

Reply:
Per our discussion Friday, my current plan is to finish addressing other issues here, land this and build per-stream instantiation on top.